giskard/rag/question_generators/ookb_questions.py

```diff
@@ -0,0 +1,176 @@
+)
+
+QUESTION_FILTER_OUTPUT_EXAMPLE = """{
+    "can_be_answered": false,
+    "correctness_reason": "The name of the Boston investment bank where Jessica worked is not directly mentioned in the context."
+}
+"""
+
+DUMMY_ANSWER = "This question can not be answered by the context. No sufficient information is provided in the context to answer this question."
+
+
+class OutOfKnowledgeBaseGenerator(GenerateFromSingleQuestionMixin, _LLMBasedQuestionGenerator):
+    """
+    Out of Knowledge Base question generator that generates questions from a KnowledgeBase.
+
+    Parameters
+    ----------
+    context_neighbors: int, optional
+        Number of context neighbors to use for question generation.
+    context_similarity_threshold: float, optional
+        Similarity threshold to keep neighboring document during question generation.
+    context_window_length: int, optional
+        Context window length of the llm used in the `llm_client` of the generator.
+    llm_client: LLMClient, optional
+        The LLM client to use for question generation. If not specified, a default openai client will be used.
+    llm_temperature: float, optional
+        The temperature to use in the LLM for question generation. The default is 0.5.
+    """
+
+    _OOKB_question_generation_prompt = QAGenerationPrompt(
+        system_prompt=OOKB_PROMPT,
+        example_input=OOKB_QUESTION_EXAMPLE_INPUT,
+        example_output=OOKB_QUESTION_EXAMPLE_OUTPUT,
+    )
+
+    _OOKB_question_check_prompt = QAGenerationPrompt(
+        system_prompt=QUESTION_CHECK_PROMPT,
+        user_input_template=QUESTION_FILTER_INPUT_TEMPLATE,
+        example_input=QUESTION_FILTER_INPUT_EXAMPLE,
+        example_output=QUESTION_FILTER_OUTPUT_EXAMPLE,
+    )
+
+    _question_type = "out of scope"
+
+    def generate_single_question(self, knowledge_base: KnowledgeBase, agent_description: str, language: str) -> dict:
+        """
+        Generate a question from a list of context documents.
+
+        Parameters
+        ----------
+        knowledge_base: KnowledgeBase
+            The knowledge base to generate the question from.
+        agent_description: str
+            The description of the agent to generate the question for.
+        language: str
+            The language to generate the question in.
+
+
+        Returns
+        -------
+        Tuple[dict, dict]
+            The generated question and the metadata of the question.
+        """
+        seed_document = knowledge_base.get_random_document()
+        context_documents = knowledge_base.get_neighbors(
+            seed_document, self._context_neighbors, self._context_similarity_threshold
+        )
+        context_str = "\n------\n".join(["", *[doc.content for doc in context_documents], ""])
+
+        reference_context = "\n\n".join([f"Document {doc.id}: {doc.content}" for doc in context_documents])
+
+        for _ in range(3):
```

Comment:
I would remove this retry loop, instead you could use retry decorator, but I would rather improve the prompt so that we don't need the retry at all.

Reply:
I am not sure whether we could achieve that but some prompt engineering is worth to do. Since current approach is quiet unstable and I have two potential improvements but not implemented and tested yet. For me I prefer we could move this retry to where this generate function be called (line 87 `giskard/rag/base.py`) instead of modifying the loop to retry decorator.

Reply:
From the experiment I did before, the success rate is about 50%-60%. I could fine one or two good examples for the prompt but not sure about removing the check part.

Reply:
If you can get to around 80% of correct questions simply by tweaking the prompt and improving the examples, it would be sufficient. The extra check call could then be removed.